[Fixes #9799] Thesaurus selectbox performance in metadata editor is very slow

mwallschlaeger opened the pull request:
References: #9799

As described in the issue. The requests to the database are very slow. This is basically related to a for loop used in the current implementation. To evaluate my changes i wrote a small benchmark to show the performance improvements done by this commit.

```
import argparse
import time

import django
django.setup()

from geonode.base.models import ThesaurusKeyword, ThesaurusKeywordLabel

parser = argparse.ArgumentParser(description='Test thesaurus performance in GeoNode')
parser.add_argument('-tid', dest='thesaurus_id', type=int, required=True, help="pass the thesaurus id from the Thesaurus Table to select catalog to run on")
parser.add_argument('-lang', dest='language', default="en", type=str, help="pass language the results are searched for")
args = parser.parse_args()

start_time = time.time()
print("starting run with current GeoNode implementation ...")
qs_local = []
qs_non_local = [("", "------")]
for key in ThesaurusKeyword.objects.filter(thesaurus_id=args.thesaurus_id):
    label = ThesaurusKeywordLabel.objects.filter(keyword=key).filter(lang=args.language)
    if label.exists():
        qs_local.append((label.get().keyword.id, label.get().label))
    else:
        qs_non_local.append((key.id, key.alt_label))

qs = qs_local + qs_non_local
print("run finished in: {} and found {} objects ...".format(round(time.time() - start_time, 5), len(qs)))

print()

start_time = time.time()
print("starting run with suggested improvement ...")

keyword_id_for_given_thesaurus = ThesaurusKeyword.objects.filter(thesaurus_id=args.thesaurus_id)  
qs_keyword_ids = ThesaurusKeywordLabel.objects.filter(lang=args.language, keyword_id__in=keyword_id_for_given_thesaurus).values("keyword_id")
not_qs_ids = ThesaurusKeywordLabel.objects.exclude(keyword_id__in=qs_keyword_ids).order_by("keyword_id").distinct("keyword_id").values("keyword_id")

qs_local = list(ThesaurusKeywordLabel.objects.filter(lang=args.language, keyword_id__in=keyword_id_for_given_thesaurus).values_list("keyword_id", "label"))
qs_non_local = list(keyword_id_for_given_thesaurus.filter(id__in=not_qs_ids).values_list("id", "alt_label"))

qs=qs_local + [("", "-------")] + qs_non_local
print("run finished in: {} and found {} objects ...".format(round(time.time() - start_time, 5), len(qs)))
```

When i run the script with the agrovoc loaded into the database (https://github.com/zalf-rdm/geonode-agrovoc-importer) on my Desktop PC i get the following results:

```
❯ python debug_test.py -tid 13
starting run with current GeoNode implementation ...
run finished in: 112.2935 and found 40251 objects ...

starting run with suggested improvement ...
run finished in: 0.26096 and found 40251 objects ...
```
So the new implementation is something like 400 times faster than the current implementation.

In my opinion the way of handling alt_labels so labels which are not available in the current language could also be improved.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/GeoNode/geonode/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in the documentation)
- [ ] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for contrib modules):
- [x] There is a ticket in https://github.com/GeoNode/geonode/issues describing the issue/improvement/feature (a notable exemption is, changes not visible to end-users)
- [x] The issue connected to the PR must have Labels and Milestone assigned
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[Fixes #<issue_number>] Title of the Issue"
- [ ] New unit tests have been added covering the changes, unless there is an explanation on why the tests are not necessary/implemented
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the QA checks: flake8 geonode
- [ ] Commits changing the **settings**, **UI**, **existing user workflows**, or adding **new functionality**, need to include documentation updates
- [ ] Commits adding **new texts** do use gettext and have updated .po / .mo files (without location infos)

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

mattiagiupponi commented on geonode/base/forms.py:
```diff
@@ -337,16 +337,14 @@ def _define_choicefield(self, item, required, tname, lang):
+        qs_non_local = list(keyword_id_for_given_thesaurus.filter(id__in=not_qs_ids).values_list("id", "alt_label"))
 
-        return qs_non_local + qs_local
+        return qs_local + [("", "-------")] + qs_non_local
```
Hi @mwallschlaeger I guess here is better to maint the `[("", "-------")]` as first, so during the preview in the select, is at the beginning. Otherwise, I guess it will be in the middle of the list.

mwallschlaeger replied:
yes and no. I think its good to have the [("", "-------")] at first. But I would suggest to put it in between the qs_local and the qs_non_local as well, so the user knows that there is a reason why there are appearing entries in another language. 

Again for my usecase with the AGROVOC there are a lot of entries not available in all languages. That means that even if a users searches for something there are always many many entries which are not related to the search.

Imagining looking for something like "v" in a list of car manufactures. Therefore I would suggest an output like:
```
---------
Vandenbrink
Vauxhall
Vector
Vencer
Venturi
Venucia
VinFast
VLF
Volkswagen
Volvo
---------
北京汽车制造厂有限公司
绅宝汽车
银翔
金杯
```

So in this case the second "---------" separates the translated entries from the not translatable ones.

mattiagiupponi replied:
Yep having it in both places is a good option

mattiagiupponi reviewed (changes requested):
Hi, @mwallschlaeger I see that no tests are added, the old ones are still valid? Or do they need some improvement?
(I'm asking since the build was failing :) )

mwallschlaeger commented:
i got your point on this. I think its good to have the "------" in the beginning of the list. But in some cases e.g. in the case of the AGROVOC you have a huge list of items not available in every language. Therefore I suggest adding a "------" in between the qs_local entries and the qs_non_local entries so users can determine where the local entries end. 

Imagine a list of car manufactures we search in with "v". But some manufactures have no translation for your language. The list would look something like:
```
-------
Vandenbrink
Vauxhall
Vector
Vencer
Venturi
Venucia
VinFast
VLF
Volkswagen
Volvo
-------
比亚迪汽车
上汽通用五菱汽车股份有限公司
奇瑞汽車股份有限公司
廣州汽車集團股份有限公司
東風汽車集團有限公司
```

In my opinion its good to show the qs_non_locals last and separate them by another "-------" entry.

mattiagiupponi commented on geonode/base/forms.py:
```diff
@@ -337,16 +337,14 @@ def _define_choicefield(self, item, required, tname, lang):
-            else:
-                qs_non_local.append((key.id, key.alt_label))
+        keyword_id_for_given_thesaurus = ThesaurusKeyword.objects.filter(thesaurus_id=item)
+        qs_keyword_ids = ThesaurusKeywordLabel.objects.filter(lang=lang, keyword_id__in=keyword_id_for_given_thesaurus).values("keyword_id")
```
Hi @mwallschlaeger 
It looks like there is some issue with changing the language. I'm trying it on master.
I'm trying it in Italian to see how it behaves. Setting in Italian, I have: `lang = it-it`,
![image](https://user-images.githubusercontent.com/51856725/182814280-9c077105-8360-4023-90b5-28ec52befe7f.png)

 so when it looks for the qs_keyword_ids, the queryset is always empty
![image](https://user-images.githubusercontent.com/51856725/182814575-68f5470b-43f4-4730-a326-0af39d7652fd.png)
 since it is saved with `lang=it`
![image](https://user-images.githubusercontent.com/51856725/182814459-d617ed1a-418a-4fe0-8539-0f5ecf8967c1.png)

So at the end qs_local is empty, 
![image](https://user-images.githubusercontent.com/51856725/182815171-d36747d1-7b36-42b3-ae7c-c0d467d9ad13.png)
while qs_non_local is full:
![image](https://user-images.githubusercontent.com/51856725/182815276-7e4dae6b-64dc-4d5e-9e32-8e56fe557dc1.png)

mattiagiupponi commented on geonode/base/views.py:
```diff
@@ -301,23 +302,28 @@ def get_results(self, context):
-        return qs_non_local + qs_local
+
+        # help to get real language string
+        requested_lang = get_language_from_request(self.request)
```
If you have time, can you please add a couple of tests on the APi?

mwallschlaeger replied:
i updated this mechanism again to be more clean in my opinion.

mwallschlaeger replied:
i am going to add a testcase for different language codes here.

mwallschlaeger commented:
OOOK. The main problem with the languages is that in django shell in django.utils.translation.get_language() returns "en" but if django.utils.translation.get_language() got executed via the browser you get a language including the country like "en-us". Therefore, i added a function which removes the country code. Sadly there is no function in the django translation module which reliable returns the language without country code. So the search functions in geonode.base.view and geonode.base.form now try to first find entries with language code including the country, because for some languages this makes sense. If no entries are found for this, we change the language to only include the language ("en"). And if this results still contain nothing geonode shows the alt_labels.

I also added some tests which set the language with and without country code and try to get results from the _get_thesauro_keyword_label function in ThesaurusAvailableForm.

In current implementation geonode always returns alt_labels because its not able to handle the language_code including country code ...

mattiagiupponi reviewed (changes requested):
overall looks ok, there is just an issue with the flake8 formatting. Can you please fix it so the CircleCI test will run?
